data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataEncoder.java

```diff
@@ -0,0 +1,340 @@
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+package com.linkedin.data.codec.entitystream;
+
+import com.linkedin.data.ByteString;
+import com.linkedin.data.Data;
+import com.linkedin.data.DataComplex;
+import com.linkedin.data.DataList;
+import com.linkedin.data.DataMap;
+import com.linkedin.entitystream.WriteHandle;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Abstract data type encoder for a {@link com.linkedin.data.DataComplex} object implemented as a
+ * {@link com.linkedin.entitystream.Writer} writing to an {@link com.linkedin.entitystream.EntityStream} of
+ * {@link ByteString}. The implementation writes to an internal non-blocking <code>OutputStream</code>
+ * implementation that has a fixed-size primary buffer and an unbounded overflow buffer. Because the bytes are pulled
+ * from the encoder asynchronously, it needs to keep the state in a stack.
+ *
+ * @author kramgopa, xma
+ */
+abstract class AbstractDataEncoder implements DataEncoder
+{
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractDataEncoder.class);
+
+  private static final Object MAP = new Object();
+  private static final Object LIST = new Object();
+
+  private Data.TraverseCallback _traverseCallback;
+  private QueueBufferedOutputStream _out;
+  private Deque<DataComplex> _stack;
+  private Deque<Iterator<?>> _iteratorStack;
+  private Deque<Object> _typeStack;
+  private WriteHandle<? super ByteString> _writeHandle;
+  private boolean _done;
+
+  private AbstractDataEncoder(int bufferSize)
+  {
+    _out = new QueueBufferedOutputStream(bufferSize);
+    _stack = new ArrayDeque<>();
+    _iteratorStack = new ArrayDeque<>();
+    _typeStack = new ArrayDeque<>();
+    _done = false;
+  }
+
+  protected AbstractDataEncoder(DataMap dataMap, int bufferSize)
+  {
+    this(bufferSize);
+
+    _stack.push(dataMap);
+    _typeStack.push(MAP);
+  }
+
+  protected AbstractDataEncoder(DataList dataList, int bufferSize)
+  {
+    this(bufferSize);
+
+    _stack.push(dataList);
+    _typeStack.push(LIST);
+  }
+
+  @Override
+  public void onInit(WriteHandle<? super ByteString> wh)
+  {
+    _writeHandle = wh;
+
+    try
+    {
+      _traverseCallback = initTraverseCallback(_out);
+    }
+    catch (IOException e)
+    {
+      _writeHandle.error(e);
+    }
+  }
+
+  abstract protected Data.TraverseCallback initTraverseCallback(OutputStream out) throws IOException;
```

Comment:
Suggest renaming this to createTraverseCallback for consistency with the blocking codec